src/H5T.c

```diff
@@ -5214,6 +5216,65 @@ H5T_path_match(H5T_path_t *path, H5T_pers_t pers, const char *name, H5T_t *src,
     FUNC_LEAVE_NOAPI(ret_value)
 } /* H5T_path_match() */
 
+/*-------------------------------------------------------------------------
+ * Function:  H5T_path_match_find_type_with_volobj
+ *
+ * Purpose:   Helper function to determine whether a datatype is or
+ *            contains a datatype that has a VOL object pointer matching
+ *            the given VOL object pointer.
+ *
+ * Return:    true/false (can't fail)
+ *
+ *-------------------------------------------------------------------------
+ */
+static bool
+H5T_path_match_find_type_with_volobj(const H5T_t *datatype, const H5VL_object_t *owned_vol_obj)
+{
+    bool ret_value = false;
+
+    FUNC_ENTER_NOAPI_NOINIT_NOERR
+
+    assert(datatype);
+    assert(owned_vol_obj);
+
+    ret_value = (datatype->shared->owned_vol_obj == owned_vol_obj);
+    if (!ret_value) {
+        switch (datatype->shared->type) {
+            case H5T_COMPOUND:
+                for (unsigned i = 0; i < datatype->shared->u.compnd.nmembs; i++) {
+                    if (ret_value)
+                        break;
+                    ret_value = H5T_path_match_find_type_with_volobj(datatype->shared->u.compnd.memb[i].type,
+                                                                     owned_vol_obj);
+                }
+                break;
+
+            case H5T_ARRAY:
+                /* Should be an error if no parent, but simplify logic for a true/false return value */
+                if (datatype->shared->parent)
+                    ret_value = H5T_path_match_find_type_with_volobj(datatype->shared->parent, owned_vol_obj);
+                break;
+
+            case H5T_INTEGER:
+            case H5T_FLOAT:
+            case H5T_TIME:
+            case H5T_STRING:
+            case H5T_BITFIELD:
+            case H5T_OPAQUE:
+            case H5T_REFERENCE: /* Should have been determined by above check */
+            case H5T_ENUM:
```

Comment:
I would think that the ENUM & VLEN parent (base) types should also be checked, why would they be covered by the earlier check?

Reply:
For enum, the base type can only be an integer type, so we shouldn't need to check that. But good point on the vlen of vlen; I wasn't thinking recursively enough!

Reply:
Is it possible that the base type of the enum is a VOL object?  (I think the answer is yes, but haven't looked in detail)

Reply:
> Is it possible that the base type of the enum is a VOL object? (I think the answer is yes, but haven't looked in detail)

I suppose with a committed datatype you might have an enum where the base type is wrapped in a VOL object, but this function is only about checking the `owned_vol_obj` field in a datatype's shared structure, which only ever gets set for variable-length and reference datatypes. Adding a check against the base type of an enum shouldn't really hurt anything, but the enum type and its base type should never have the `owned_vol_obj` field set.